File: frontend/src/api/report.ts
import request from './request'

// 报表列表项
export interface ReportListItem {
  id: number
  name: string
  reportType: string
  coverImage?: string
  isPublished: boolean
  pageCount: number
  createdAt: string
  updatedAt?: string
}

// 报表详情
export interface ReportDetail {
  id: number
  name: string
  reportType: string
  coverImage?: string
  configJson: string
  remark?: string
  isPublished: boolean
  publishedAt?: string
  createdAt: string
  updatedAt?: string
  pages: ReportPage[]
}

// 报表页面
export interface ReportPage {
  id: number
  title: string
  sortOrder: number
  configJson: string
  items: ReportItem[]
}

// 报表元素
export interface ReportItem {
  id: number
  itemType: 'chart' | 'panel' | 'text' | 'image' | 'shape'
  chartId?: number
  chartName?: string
  panelId?: number
  panelName?: string
  textContent?: string
  imageUrl?: string
  layoutJson: string
  styleJson: string
  sortOrder: number
}

// 创建报表请求
export interface ReportCreate {
  name: string
  reportType?: string
  coverImage?: string
  configJson?: string
  remark?: string
}

// 更新报表请求
export interface ReportUpdate extends ReportCreate {}

// 保存页面请求
export interface ReportPageSave {
  id?: number
  title: string
  sortOrder: number
  configJson?: string
}

// 保存元素请求
export interface ReportItemSave {
  id?: number
  itemType: string
  chartId?: number
  panelId?: number
  textContent?: string
  imageUrl?: string
  layoutJson: string
  styleJson?: string
  sortOrder: number
}

// 获取报表列表
export function getReportList() {
  return request.get<ReportListItem[]>('/reports')
}

// 获取报表详情
export function getReportDetail(id: number) {
  return request.get<ReportDetail>(`/reports/${id}`)
}

// 创建报表
export function createReport(data: ReportCreate) {
  return request.post<{ id: number }>('/reports', data)
}

// 更新报表
export function updateReport(id: number, data: ReportUpdate) {
  return request.put<void>(`/reports/${id}`, data)
}

// 删除报表
export function deleteReport(id: number) {
  return request.delete<void>(`/reports/${id}`)
}

// 保存页面
export function saveReportPage(reportId: number, data: ReportPageSave) {
  return request.post<{ id: number }>(`/reports/${reportId}/pages`, data)
}

// 删除页面
export function deleteReportPage(pageId: number) {
  return request.delete<void>(`/reports/pages/${pageId}`)
}

// 保存元素
export function saveReportItem(pageId: number, data: ReportItemSave) {
  return request.post<{ id: number }>(`/reports/pages/${pageId}/items`, data)
}

// 删除元素
export function deleteReportItem(itemId: number) {
  return request.delete<void>(`/reports/items/${itemId}`)
}

// 获取报表渲染数据
export function getReportRenderData(id: number) {
  return request.get<any>(`/reports/${id}/render`)
}

File: backend/Bi.Domain/Entities/Report.cs
namespace Bi.Domain.Entities;

/// <summary>
/// 报表/报告主表
/// 支持多页报告结构，可包含图表、面板、文字、图片等元素
/// </summary>
public class BiReport : BaseEntity
{
    /// <summary>
    /// 报表名称
    /// </summary>
    public string Name { get; set; } = string.Empty;
    
    /// <summary>
    /// 报表类型：report(报告)/dashboard(仪表板)
    /// </summary>
    public string ReportType { get; set; } = "report";
    
    /// <summary>
    /// 封面图URL
    /// </summary>
    public string? CoverImage { get; set; }
    
    /// <summary>
    /// 全局配置JSON（主题、页面尺寸等）
    /// </summary>
    public string ConfigJson { get; set; } = "{}";
    
    /// <summary>
    /// 备注说明
    /// </summary>
    public string? Remark { get; set; }
    
    /// <summary>
    /// 是否已发布
    /// </summary>
    public bool IsPublished { get; set; }
    
    /// <summary>
    /// 发布时间
    /// </summary>
    public DateTime? PublishedAt { get; set; }
    
    /// <summary>
    /// 创建人ID
    /// </summary>
    public long CreatedBy { get; set; }
    
    /// <summary>
    /// 报表页列表
    /// </summary>
    public List<BiReportPage> Pages { get; set; } = new();
}

/// <summary>
/// 报表页面
/// 每个报表可包含多个页面，类似PPT结构
/// </summary>
public class BiReportPage : BaseEntity
{
    /// <summary>
    /// 所属报表ID
    /// </summary>
    public long ReportId { get; set; }
    
    /// <summary>
    /// 页面标题
    /// </summary>
    public string Title { get; set; } = string.Empty;
    
    /// <summary>
    /// 页面排序
    /// </summary>
    public int SortOrder { get; set; }
    
    /// <summary>
    /// 页面配置JSON（背景、布局等）
    /// </summary>
    public string ConfigJson { get; set; } = "{}";
    
    /// <summary>
    /// 所属报表
    /// </summary>
    public BiReport Report { get; set; } = null!;
    
    /// <summary>
    /// 页面元素列表
    /// </summary>
    public List<BiReportItem> Items { get; set; } = new();
}

/// <summary>
/// 报表元素/组件
/// 支持多种类型：图表、面板、文本、图片、表格等
/// </summary>
public class BiReportItem : BaseEntity
{
    /// <summary>
    /// 所属页面ID
    /// </summary>
    public long PageId { get; set; }
    
    /// <summary>
    /// 元素类型：chart/panel/text/image/table/shape
    /// </summary>
    public string ItemType { get; set; } = "chart";
    
    /// <summary>
    /// 关联图表ID（当ItemType=chart时）
    /// </summary>
    public long? ChartId { get; set; }
    
    /// <summary>
    /// 关联面板ID（当ItemType=panel时）
    /// </summary>
    public long? PanelId { get; set; }
    
    /// <summary>
    /// 文本内容（当ItemType=text时）
    /// </summary>
    public string? TextContent { get; set; }
    
    /// <summary>
    /// 图片URL（当ItemType=image时）
    /// </summary>
    public string? ImageUrl { get; set; }
    
    /// <summary>
    /// 布局配置JSON（x, y, width, height, zIndex等）
    /// </summary>
    public string LayoutJson { get; set; } = "{}";
    
    /// <summary>
    /// 样式配置JSON（字体、颜色、边框等）
    /// </summary>
    public string StyleJson { get; set; } = "{}";
    
    /// <summary>
    /// 排序顺序
    /// </summary>
    public int SortOrder { get; set; }
    
    /// <summary>
    /// 所属页面
    /// </summary>
    public BiReportPage Page { get; set; } = null!;
    
    /// <summary>
    /// 关联图表
    /// </summary>
    public Chart? Chart { get; set; }

    /// <summary>
    /// 关联面板
    /// </summary>
    public Panel? Panel { get; set; }
}

